File: app/src/main/java/com/totallytim/randomreminders/modules/Schedule.java
package com.totallytim.randomreminders.modules;

import android.content.Context;

import androidx.appcompat.app.AppCompatActivity;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;

public class Schedule {

    // TODO: clean up a little

    // format of [days][hours], week starts on Sunday, 24H clock
    // for [hours]: T || F if it can be used (any minute within that hour)
    private boolean[][] week;
    private List<Reminder> reminders;

    // TODO: import JSON as schedule
    public Schedule(Context context) {
        JSONObject json;
        try {
            json = new JSONObject(
                    Schedule.loadJsonAssets(context, "schedule.json"));

            // TODO: convert JSON to valid array (or vice versa)
            
        } catch (IOException e) {
            e.printStackTrace();
        } catch (JSONException e) {
            e.printStackTrace();
        }



        throw new UnsupportedOperationException();
    }

    public Schedule(Context context, List<Reminder> reminders) {
        this(context);
        this.reminders = reminders;
    }

    private static String loadJsonAssets(Context context, String jsonPath) throws IOException {
        String json;
        try {
            InputStream stream = context.getAssets().open(jsonPath);

            int size = stream.available();
            byte[] buffer = new byte[size];

            stream.read(buffer);
            stream.close();

            json = new String(buffer, "utf-8");
        } catch(FileNotFoundException e) {
            System.out.println("No existing JSON object");
            return "";
        }
        return json;
    }

    public Schedule(boolean[][] availableTimes) {
        this.week = availableTimes;
        this.reminders = new ArrayList<>();
    }

    public Schedule(boolean[][] availableTimes, List<Reminder> reminders) {
        this(availableTimes);
        this.reminders = reminders;
    }

    public static String getDayString(int day) {
        switch(day) {
            case 0:
                return "Sunday";
            case 1:
                return "Monday";
            case 2:
                return "Tuesday";
            case 3:
                return "Wednesday";
            case 4:
                return "Thursday";
            case 5:
                return "Friday";
            case 6:
                return "Saturday";
            default:
                throw new UnsupportedOperationException(String.format("Index '%d' not recognised", day));
        }
    }

    public static int getDayIndex(String day) {
        switch(day) {
            case "Sunday":
                return 0;
            case "Monday":
                return 1;
            case "Tuesday":
                return 2;
            case "Wednesday":
                return 3;
            case "Thursday":
                return 4;
            case "Friday":
                return 5;
            case "Saturday":
                return 6;
            default:
                throw new UnsupportedOperationException(String.format("Day '%s' not recognised", day));
        }
    }

    public boolean[] getAvailableDay(String day) {
        return this.week[getDayIndex(day)];
    }

    public boolean isAvailableHour(String day, int hour) {
        return this.getAvailableDay(day)[hour];
    }

    public boolean isValidTime(Date time) {
        Calendar c = Calendar.getInstance();
        c.setTime(time);

        int day = c.get(Calendar.DAY_OF_WEEK);
        int hour = c.get(Calendar.HOUR_OF_DAY);

        return isAvailableHour(getDayString(day), hour);
    }

    // TODO: CRITICAL
    // TODO: verify that I have correct formats for times
    public void setNextReminder(Reminder reminder) {
        if(!this.reminders.contains(reminder)) {
            this.reminders.add(reminder);
        }

        long startMillis = System.currentTimeMillis();
        long endMillis = (long)(startMillis + reminder.getFrequency());

        long instanceDate = (long) (Math.random() * (endMillis - startMillis));

        reminder.setNextOccurrence(new Date(startMillis + instanceDate));
    }

    public Date getNextReminder(Reminder reminder) throws NullPointerException {
        if(!this.reminders.contains(reminder)) {
            throw new NullPointerException();
        }
        return this.reminders.get(this.reminders.indexOf(reminder)).getNextOccurrence();
    }

    public List<Reminder> getReminders() {
        return this.reminders;
    }

    public void addReminder(Reminder reminder) {
        setNextReminder(reminder);
    }

    public boolean[][] getWeek() {
        return week;
    }

    public void setWeek(boolean[][] week) {
        this.week = week;
    }

    // TODO: export as JSON to local for later import
    public void exportJson() {
        throw new UnsupportedOperationException();
    }
}

File: app/src/main/java/com/totallytim/randomreminders/modules/Schedule.kt
package com.totallytim.randomreminders.modules

import androidx.lifecycle.LiveData
import com.totallytim.randomreminders.asMutableList
import com.totallytim.randomreminders.database.Day
import com.totallytim.randomreminders.database.Reminder
import com.totallytim.randomreminders.database.ReminderDatabase
import com.totallytim.randomreminders.fromCalendar
import kotlinx.coroutines.*
import java.lang.IllegalArgumentException
import java.util.*

/**
 * The schedule for when all reminders that exist will trigger events in the main application, stores
 * values as compact hexdec time values and caches them locally, before storing them according to the
 * DAO.
 */
class Schedule(var database: ReminderDatabase) {

    // available times and reminders data
    private var week: LiveData<Array<Day>> = database.reminderDatabaseDao.getAllDays()
    private var reminders: LiveData<Array<Reminder>> = database.reminderDatabaseDao.getAllReminders()

    // scope of current tasks
    private var viewModelJob = Job()
    private val uiScope = CoroutineScope(Dispatchers.Main + viewModelJob)

    /**
     * Returns what days are available from their string name.
     *
     * @param day   The day of the week
     * @return the data object containing when any reminders may trigger
     * @throws IllegalArgumentException
     *              if an invalid day string is given
     */
    @Throws(IllegalArgumentException::class)
    fun getAvailableDay(day: String): Day {
        return when(day) {
            // asserting non-null, hopefully that's what it needs
            "Sunday" -> week.value!![0]
            "Monday" -> week.value!![1]
            "Tuesday" -> week.value!![2]
            "Wednesday" -> week.value!![3]
            "Thursday" -> week.value!![4]
            "Friday" -> week.value!![5]
            "Saturday" -> week.value!![6]
            else -> throw IllegalArgumentException("Unknown day")
        }
    }

    /**
     * Asserts that the given day string has a valid time of the HH:mm given
     *
     * @param day       the day of the week
     * @param hours     the hour queried
     * @param minutes   the minute queried
     * @return true iff the time period is valid for a reminder
     */
    fun isAvailableHour(day: String, hours: Int, minutes: Int): Boolean {
        return getAvailableDay(day).isValidTime(hours, minutes)
    }

    /**
     * Sets the next time for a Reminder event to trigger.
     *
     * @param reminder the Reminder to be triggered
     */
    fun setNextReminder(reminder: Reminder) {
        val calendar = Calendar.getInstance()
        val rand = (2 * reminder.variance) * Math.random() - reminder.variance
        calendar.add(Calendar.DAY_OF_WEEK, (reminder.frequency + rand).toInt())
//        if(reminders.value!!.contains(reminder)) {
//            reminders.value!![reminders.value!!.indexOf(reminder)].nextOccurrence = fromCalendar(calendar)
//        } else {
//            reminder.nextOccurrence = fromCalendar(calendar)
//            reminders.value!!.add(reminder)
//        }

        // TODO: update database && refresh (instance only!)
    }

    /**
     * Inserts a new reminder to the database, and sets a trigger time for it.
     *
     * @param reminder  A new reminder object
     */
    // TODO: implement replacement correctly
    @Deprecated("Moved to initialise new reminders in 'New Reminder' fragment")
    fun addReminder(reminder: Reminder) {
        setNextReminder(reminder)

        // TODO: insert to database
    }

    /**
     * Retrieves the next reminder to be triggered
     *
     * @return the next reminder by calendar date
     *
     */
    fun getNextReminder(): Reminder? {
        return if(reminders.value!!.isEmpty()) {
            null
        } else {
            val sortedList = reminders.value!!.sortedWith(compareBy { it.nextOccurrence })
            return sortedList[0]
        }
    }

    /**
     * Refreshes all data after knowledge of database update.
     *
     * @return context of updated data entries
     */
    private suspend fun refreshData() {
        // TODO: change scope
//        return withContext()
        uiScope.launch {
            // TODO: implement refresh via keyed changes for better efficiency
            week = database.reminderDatabaseDao.getAllDays()
            reminders = database.reminderDatabaseDao.getAllReminders()
        }
    }
}